{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7666491b-8d3f-4b39-be9d-6cfb602e3b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Install abagen using most up-to-date version from github\n",
    "# !git clone https://github.com/rmarkello/abagen.git\n",
    "# !pip install ./abagen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "309b1d88-e3a1-4c49-a1c5-8f4246f4f252",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load abagen toolbox\n",
    "import abagen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0f40d2d-b833-4ea4-a390-062a0c0c2da3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download AHBA microarray data\n",
    "files = abagen.fetch_microarray(donors='all',data_dir=\"/home/jovyan/shared/abagen-data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "09fbb62b-bd10-46ac-b8aa-b6e7688476a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'9861': {'microarray': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor9861/MicroarrayExpression.csv',\n",
       "  'ontology': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor9861/Ontology.csv',\n",
       "  'pacall': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor9861/PACall.csv',\n",
       "  'probes': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor9861/Probes.csv',\n",
       "  'annotation': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor9861/SampleAnnot.csv'},\n",
       " '10021': {'microarray': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor10021/MicroarrayExpression.csv',\n",
       "  'ontology': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor10021/Ontology.csv',\n",
       "  'pacall': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor10021/PACall.csv',\n",
       "  'probes': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor10021/Probes.csv',\n",
       "  'annotation': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor10021/SampleAnnot.csv'},\n",
       " '12876': {'microarray': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor12876/MicroarrayExpression.csv',\n",
       "  'ontology': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor12876/Ontology.csv',\n",
       "  'pacall': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor12876/PACall.csv',\n",
       "  'probes': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor12876/Probes.csv',\n",
       "  'annotation': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor12876/SampleAnnot.csv'},\n",
       " '14380': {'microarray': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor14380/MicroarrayExpression.csv',\n",
       "  'ontology': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor14380/Ontology.csv',\n",
       "  'pacall': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor14380/PACall.csv',\n",
       "  'probes': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor14380/Probes.csv',\n",
       "  'annotation': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor14380/SampleAnnot.csv'},\n",
       " '15496': {'microarray': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15496/MicroarrayExpression.csv',\n",
       "  'ontology': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15496/Ontology.csv',\n",
       "  'pacall': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15496/PACall.csv',\n",
       "  'probes': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15496/Probes.csv',\n",
       "  'annotation': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15496/SampleAnnot.csv'},\n",
       " '15697': {'microarray': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15697/MicroarrayExpression.csv',\n",
       "  'ontology': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15697/Ontology.csv',\n",
       "  'pacall': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15697/PACall.csv',\n",
       "  'probes': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15697/Probes.csv',\n",
       "  'annotation': '/home/jovyan/shared/abagen-data/microarray/normalized_microarray_donor15697/SampleAnnot.csv'}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Print to screen files object to check what's there, should show gene expression data .csv files for 6 donors\n",
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9fe80e52-f920-4f4f-ad0f-9b32e9dbd023",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/data/atlas-desikankilliany-lh.label.gii.gz',\n",
       " '/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/data/atlas-desikankilliany-rh.label.gii.gz')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Specify atlas for abagen to use to be surface version of Desikan-Killiany atlas and view names of files specified for use\n",
    "atlas = abagen.fetch_desikan_killiany(surface=True)\n",
    "atlas['image']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "369019bb-ab3d-4b90-bffb-06cbbfb928b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/probes_.py:147: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  return io.read_microarray(microarray).groupby(sid, axis=1).mean()\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/probes_.py:147: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  return io.read_microarray(microarray).groupby(sid, axis=1).mean()\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/probes_.py:147: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  return io.read_microarray(microarray).groupby(sid, axis=1).mean()\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/probes_.py:147: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  return io.read_microarray(microarray).groupby(sid, axis=1).mean()\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/probes_.py:147: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  return io.read_microarray(microarray).groupby(sid, axis=1).mean()\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/probes_.py:147: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  return io.read_microarray(microarray).groupby(sid, axis=1).mean()\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:405: FutureWarning: The provided callable <function mean at 0x7f58f0fd76d0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  microarray.groupby('label').aggregate(metric),\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:404: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  gene_by_label = pd.concat([\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:405: FutureWarning: The provided callable <function mean at 0x7f58f0fd76d0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  microarray.groupby('label').aggregate(metric),\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:404: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  gene_by_label = pd.concat([\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:405: FutureWarning: The provided callable <function mean at 0x7f58f0fd76d0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  microarray.groupby('label').aggregate(metric),\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:404: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  gene_by_label = pd.concat([\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:405: FutureWarning: The provided callable <function mean at 0x7f58f0fd76d0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  microarray.groupby('label').aggregate(metric),\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:404: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  gene_by_label = pd.concat([\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:405: FutureWarning: The provided callable <function mean at 0x7f58f0fd76d0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  microarray.groupby('label').aggregate(metric),\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:404: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  gene_by_label = pd.concat([\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:405: FutureWarning: The provided callable <function mean at 0x7f58f0fd76d0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  microarray.groupby('label').aggregate(metric),\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:404: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  gene_by_label = pd.concat([\n",
      "/srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/samples_.py:481: FutureWarning: The provided callable <function mean at 0x7f58f0fd76d0> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  microarray = pd.concat(microarray).groupby('label').aggregate(metric)\n"
     ]
    }
   ],
   "source": [
    "#Parcellate expression data into surface version of Desikan-Killiany Atlas\n",
    "expression = abagen.get_expression_data(atlas['image'], data_dir=\"/home/jovyan/shared/abagen-data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "31189fd3-1a9d-4bd7-b01b-385f9d507f6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_symbol</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADACL3</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.440726</td>\n",
       "      <td>0.589278</td>\n",
       "      <td>0.381510</td>\n",
       "      <td>0.436797</td>\n",
       "      <td>0.467330</td>\n",
       "      <td>0.483065</td>\n",
       "      <td>0.452128</td>\n",
       "      <td>0.543946</td>\n",
       "      <td>0.536256</td>\n",
       "      <td>0.588966</td>\n",
       "      <td>...</td>\n",
       "      <td>0.486024</td>\n",
       "      <td>0.533246</td>\n",
       "      <td>0.522263</td>\n",
       "      <td>0.601487</td>\n",
       "      <td>0.587398</td>\n",
       "      <td>0.405509</td>\n",
       "      <td>0.634875</td>\n",
       "      <td>0.497207</td>\n",
       "      <td>0.666547</td>\n",
       "      <td>0.477138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.678336</td>\n",
       "      <td>0.431307</td>\n",
       "      <td>0.572806</td>\n",
       "      <td>0.671694</td>\n",
       "      <td>0.480824</td>\n",
       "      <td>0.503408</td>\n",
       "      <td>0.355954</td>\n",
       "      <td>0.368647</td>\n",
       "      <td>0.556326</td>\n",
       "      <td>0.584441</td>\n",
       "      <td>...</td>\n",
       "      <td>0.387346</td>\n",
       "      <td>0.363805</td>\n",
       "      <td>0.343892</td>\n",
       "      <td>0.549072</td>\n",
       "      <td>0.387021</td>\n",
       "      <td>0.633337</td>\n",
       "      <td>0.349052</td>\n",
       "      <td>0.223801</td>\n",
       "      <td>0.406646</td>\n",
       "      <td>0.503784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.520608</td>\n",
       "      <td>0.446375</td>\n",
       "      <td>0.609580</td>\n",
       "      <td>0.459314</td>\n",
       "      <td>0.501768</td>\n",
       "      <td>0.542337</td>\n",
       "      <td>0.516451</td>\n",
       "      <td>0.527325</td>\n",
       "      <td>0.416930</td>\n",
       "      <td>0.433428</td>\n",
       "      <td>...</td>\n",
       "      <td>0.486235</td>\n",
       "      <td>0.506752</td>\n",
       "      <td>0.408213</td>\n",
       "      <td>0.448193</td>\n",
       "      <td>0.453520</td>\n",
       "      <td>0.438250</td>\n",
       "      <td>0.429106</td>\n",
       "      <td>0.587044</td>\n",
       "      <td>0.404144</td>\n",
       "      <td>0.529791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.317040</td>\n",
       "      <td>0.307230</td>\n",
       "      <td>0.451889</td>\n",
       "      <td>0.477776</td>\n",
       "      <td>0.474542</td>\n",
       "      <td>0.471519</td>\n",
       "      <td>0.687771</td>\n",
       "      <td>0.399432</td>\n",
       "      <td>0.566165</td>\n",
       "      <td>0.440505</td>\n",
       "      <td>...</td>\n",
       "      <td>0.590275</td>\n",
       "      <td>0.526024</td>\n",
       "      <td>0.370078</td>\n",
       "      <td>0.550802</td>\n",
       "      <td>0.423469</td>\n",
       "      <td>0.451572</td>\n",
       "      <td>0.444684</td>\n",
       "      <td>0.829026</td>\n",
       "      <td>0.733465</td>\n",
       "      <td>0.387726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.711983</td>\n",
       "      <td>0.261514</td>\n",
       "      <td>0.699565</td>\n",
       "      <td>0.514992</td>\n",
       "      <td>0.455938</td>\n",
       "      <td>0.476981</td>\n",
       "      <td>0.313874</td>\n",
       "      <td>0.296195</td>\n",
       "      <td>0.600913</td>\n",
       "      <td>0.489880</td>\n",
       "      <td>...</td>\n",
       "      <td>0.368767</td>\n",
       "      <td>0.372662</td>\n",
       "      <td>0.779778</td>\n",
       "      <td>0.435531</td>\n",
       "      <td>0.376815</td>\n",
       "      <td>0.562862</td>\n",
       "      <td>0.542200</td>\n",
       "      <td>0.085269</td>\n",
       "      <td>0.453389</td>\n",
       "      <td>0.591145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>0.255569</td>\n",
       "      <td>0.460440</td>\n",
       "      <td>0.616346</td>\n",
       "      <td>0.509861</td>\n",
       "      <td>0.452747</td>\n",
       "      <td>0.365555</td>\n",
       "      <td>0.513271</td>\n",
       "      <td>0.641262</td>\n",
       "      <td>0.443922</td>\n",
       "      <td>0.705150</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693676</td>\n",
       "      <td>0.467304</td>\n",
       "      <td>0.524012</td>\n",
       "      <td>0.511560</td>\n",
       "      <td>0.485877</td>\n",
       "      <td>0.508755</td>\n",
       "      <td>0.604058</td>\n",
       "      <td>0.630165</td>\n",
       "      <td>0.742505</td>\n",
       "      <td>0.400399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>0.642150</td>\n",
       "      <td>0.713728</td>\n",
       "      <td>0.140265</td>\n",
       "      <td>0.354118</td>\n",
       "      <td>0.381271</td>\n",
       "      <td>0.193237</td>\n",
       "      <td>0.225932</td>\n",
       "      <td>0.393234</td>\n",
       "      <td>0.810565</td>\n",
       "      <td>0.335436</td>\n",
       "      <td>...</td>\n",
       "      <td>0.357386</td>\n",
       "      <td>0.772246</td>\n",
       "      <td>0.674840</td>\n",
       "      <td>0.980960</td>\n",
       "      <td>0.358707</td>\n",
       "      <td>0.262772</td>\n",
       "      <td>0.575014</td>\n",
       "      <td>0.311585</td>\n",
       "      <td>0.050366</td>\n",
       "      <td>0.679396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>0.874019</td>\n",
       "      <td>0.809826</td>\n",
       "      <td>0.938905</td>\n",
       "      <td>0.087033</td>\n",
       "      <td>0.027779</td>\n",
       "      <td>0.749595</td>\n",
       "      <td>0.780226</td>\n",
       "      <td>0.845095</td>\n",
       "      <td>0.253037</td>\n",
       "      <td>0.485441</td>\n",
       "      <td>...</td>\n",
       "      <td>0.605505</td>\n",
       "      <td>0.698701</td>\n",
       "      <td>0.812223</td>\n",
       "      <td>0.236611</td>\n",
       "      <td>0.085663</td>\n",
       "      <td>0.505108</td>\n",
       "      <td>0.586400</td>\n",
       "      <td>0.433677</td>\n",
       "      <td>0.538252</td>\n",
       "      <td>0.523899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>0.575880</td>\n",
       "      <td>0.257826</td>\n",
       "      <td>0.764119</td>\n",
       "      <td>0.981280</td>\n",
       "      <td>0.818433</td>\n",
       "      <td>0.382591</td>\n",
       "      <td>0.787600</td>\n",
       "      <td>0.344107</td>\n",
       "      <td>0.567647</td>\n",
       "      <td>0.447769</td>\n",
       "      <td>...</td>\n",
       "      <td>0.140754</td>\n",
       "      <td>0.141557</td>\n",
       "      <td>0.266908</td>\n",
       "      <td>0.447734</td>\n",
       "      <td>0.212665</td>\n",
       "      <td>0.817909</td>\n",
       "      <td>0.444072</td>\n",
       "      <td>0.768432</td>\n",
       "      <td>0.605523</td>\n",
       "      <td>0.300819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>0.698525</td>\n",
       "      <td>0.568063</td>\n",
       "      <td>0.679651</td>\n",
       "      <td>0.522676</td>\n",
       "      <td>0.423293</td>\n",
       "      <td>0.445324</td>\n",
       "      <td>0.436730</td>\n",
       "      <td>0.536829</td>\n",
       "      <td>0.495981</td>\n",
       "      <td>0.518441</td>\n",
       "      <td>...</td>\n",
       "      <td>0.421602</td>\n",
       "      <td>0.553703</td>\n",
       "      <td>0.535135</td>\n",
       "      <td>0.595204</td>\n",
       "      <td>0.477346</td>\n",
       "      <td>0.459767</td>\n",
       "      <td>0.414425</td>\n",
       "      <td>0.344454</td>\n",
       "      <td>0.395037</td>\n",
       "      <td>0.592096</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 15633 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_symbol      A1BG  A1BG-AS1       A2M     A2ML1   A3GALT2    A4GALT  \\\n",
       "label                                                                     \n",
       "1            0.440726  0.589278  0.381510  0.436797  0.467330  0.483065   \n",
       "2            0.678336  0.431307  0.572806  0.671694  0.480824  0.503408   \n",
       "3            0.520608  0.446375  0.609580  0.459314  0.501768  0.542337   \n",
       "4            0.317040  0.307230  0.451889  0.477776  0.474542  0.471519   \n",
       "5            0.711983  0.261514  0.699565  0.514992  0.455938  0.476981   \n",
       "...               ...       ...       ...       ...       ...       ...   \n",
       "71           0.255569  0.460440  0.616346  0.509861  0.452747  0.365555   \n",
       "72           0.642150  0.713728  0.140265  0.354118  0.381271  0.193237   \n",
       "73           0.874019  0.809826  0.938905  0.087033  0.027779  0.749595   \n",
       "74           0.575880  0.257826  0.764119  0.981280  0.818433  0.382591   \n",
       "75           0.698525  0.568063  0.679651  0.522676  0.423293  0.445324   \n",
       "\n",
       "gene_symbol      AAAS      AACS   AADACL3     AADAT  ...      ZW10    ZWILCH  \\\n",
       "label                                                ...                       \n",
       "1            0.452128  0.543946  0.536256  0.588966  ...  0.486024  0.533246   \n",
       "2            0.355954  0.368647  0.556326  0.584441  ...  0.387346  0.363805   \n",
       "3            0.516451  0.527325  0.416930  0.433428  ...  0.486235  0.506752   \n",
       "4            0.687771  0.399432  0.566165  0.440505  ...  0.590275  0.526024   \n",
       "5            0.313874  0.296195  0.600913  0.489880  ...  0.368767  0.372662   \n",
       "...               ...       ...       ...       ...  ...       ...       ...   \n",
       "71           0.513271  0.641262  0.443922  0.705150  ...  0.693676  0.467304   \n",
       "72           0.225932  0.393234  0.810565  0.335436  ...  0.357386  0.772246   \n",
       "73           0.780226  0.845095  0.253037  0.485441  ...  0.605505  0.698701   \n",
       "74           0.787600  0.344107  0.567647  0.447769  ...  0.140754  0.141557   \n",
       "75           0.436730  0.536829  0.495981  0.518441  ...  0.421602  0.553703   \n",
       "\n",
       "gene_symbol     ZWINT      ZXDA      ZXDB      ZXDC    ZYG11B       ZYX  \\\n",
       "label                                                                     \n",
       "1            0.522263  0.601487  0.587398  0.405509  0.634875  0.497207   \n",
       "2            0.343892  0.549072  0.387021  0.633337  0.349052  0.223801   \n",
       "3            0.408213  0.448193  0.453520  0.438250  0.429106  0.587044   \n",
       "4            0.370078  0.550802  0.423469  0.451572  0.444684  0.829026   \n",
       "5            0.779778  0.435531  0.376815  0.562862  0.542200  0.085269   \n",
       "...               ...       ...       ...       ...       ...       ...   \n",
       "71           0.524012  0.511560  0.485877  0.508755  0.604058  0.630165   \n",
       "72           0.674840  0.980960  0.358707  0.262772  0.575014  0.311585   \n",
       "73           0.812223  0.236611  0.085663  0.505108  0.586400  0.433677   \n",
       "74           0.266908  0.447734  0.212665  0.817909  0.444072  0.768432   \n",
       "75           0.535135  0.595204  0.477346  0.459767  0.414425  0.344454   \n",
       "\n",
       "gene_symbol     ZZEF1      ZZZ3  \n",
       "label                            \n",
       "1            0.666547  0.477138  \n",
       "2            0.406646  0.503784  \n",
       "3            0.404144  0.529791  \n",
       "4            0.733465  0.387726  \n",
       "5            0.453389  0.591145  \n",
       "...               ...       ...  \n",
       "71           0.742505  0.400399  \n",
       "72           0.050366  0.679396  \n",
       "73           0.538252  0.523899  \n",
       "74           0.605523  0.300819  \n",
       "75           0.395037  0.592096  \n",
       "\n",
       "[68 rows x 15633 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Print to screen contents of newly generated expression (panda.dataframe) object, should show regions as rows by gene as columns\n",
    "expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "84b398fb-927e-44e2-a154-65bf64bcf454",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate report of what parameters were used to process and parcellate microarray data\n",
    "from abagen import reporting\n",
    "generator = reporting.Report(atlas['image'], atlas['info'])\n",
    "report = generator.gen_report()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "62f97d8c-41f1-4c17-b1b8-37be0bcef2d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Regional microarry expression data were obtained from 6 post-mortem brains (1 female, ages 24.0--57.0, 42.50 +/- 13.38) provided by the Allen Human Brain Atlas (AHBA, https://human.brain-map.org; [H2012N]). Data were processed with the abagen toolbox (version 0.1.4+15.gdc4a007; https://github.com/rmarkello/abagen; [M2021B]) using a 68-region surface-based atlas in MNI space.\\n\\nFirst, microarray probes were reannotated using data provided by [A2019N]; probes not matched to a valid Entrez ID were discarded. Next, probes were filtered based on their expression intensity relative to background noise [Q2002N], such that probes with intensity less than the background in >=50.00% of samples across donors were discarded . When multiple probes indexed the expression of the same gene, we selected and used the probe with the most consistent pattern of regional variation across donors (i.e., differential stability; [H2015N]), calculated with:\\n\\n$$ \\\\Delta_{{S}}(p) = \\\\frac{{1}}{{\\\\binom{{N}}{{2}}}} \\\\, \\\\sum_{{i=1}}^{{N-1}} \\\\sum_{{j=i+1}}^{{N}} \\\\rho[B_{{i}}(p), B_{{j}}(p)] $$\\n\\nwhere $ \\\\rho $ is Spearman's rank correlation of the expression of a single probe, p, across regions in two donors $B_{{i}}$ and $B_{{j}}$, and N is the total number of donors. Here, regions correspond to the structural designations provided in the ontology from the AHBA. \\n\\nThe MNI coordinates of tissue samples were updated to those generated via non-linear registration using the Advanced Normalization Tools (ANTs; https://github.com/chrisfilo/alleninf). Samples were assigned to brain regions by minimizing the Euclidean distance between the MNI coordinates of each sample and the nearest surface vertex. Samples where the Euclidean distance to the nearest vertex was more than 2 standard deviations above the mean distance for all samples belonging to that donor were excluded. To reduce the potential for misassignment, sample-to-region matching was constrained by hemisphere and gross structural divisions (i.e., cortex, subcortex/brainstem, and cerebellum, such that e.g., a sample in the left cortex could only be assigned to an atlas parcel in the left cortex; [A2019N]). All tissue samples not assigned to a brain region in the provided atlas were discarded. \\n\\nInter-subject variation was addressed by normalizing tissue sample expression values across genes using a robust sigmoid function [F2013J]:\\n\\n$$ x_{{norm}} = \\\\frac{{1}}{{1 + \\\\exp(-\\\\frac{{(x-\\\\langle x \\\\rangle)}} {{\\\\text{{IQR}}_{{x}}}})}} $$\\n\\nwhere $\\\\langle x \\\\rangle$ is the median and $\\\\text{{IQR}}_{{x}}$ is the normalized interquartile range of the expression of a single tissue sample across genes. Normalized expression values were then rescaled to the unit interval: \\n\\n$$ x_{{scaled}} = \\\\frac{{x_{{norm}} - \\\\min(x_{{norm}})}} {{\\\\max(x_{{norm}}) - \\\\min(x_{{norm}})}} $$\\n\\nGene expression values were then normalized across tissue samples using an identical procedure. Samples assigned to the same brain region were averaged separately for each donor and then across donors, yielding a regional expression matrix \\x08.\\n\\nREFERENCES\\n----------\\n[A2019N]: Arnatkevic̆iūtė, A., Fulcher, B. D., & Fornito, A. (2019). A practical guide to linking brain-wide gene expression and neuroimaging data. Neuroimage, 189, 353-367.\\n[F2013J]: Fulcher, B. D., Little, M. A., & Jones, N. S. (2013). Highly comparative time-series analysis: the empirical structure of time series and their methods. Journal of the Royal Society Interface, 10(83), 20130048.\\n[H2012N]: Hawrylycz, M. J., Lein, E. S., Guillozet-Bongaarts, A. L., Shen, E. H., Ng, L., Miller, J. A., ... & Jones, A. R. (2012). An anatomically comprehensive atlas of the adult human brain transcriptome. Nature, 489(7416), 391-399.\\n[H2015N]: Hawrylycz, M., Miller, J. A., Menon, V., Feng, D., Dolbeare, T., Guillozet-Bongaarts, A. L., ... & Lein, E. (2015). Canonical genetic signatures of the adult human brain. Nature Neuroscience, 18(12), 1832.\\n[M2021B]: Markello, R.D., Arnatkevic̆iūtė, A., Poline, J-B., Fulcher, B. D., Fornito, A., & Misic, B. (2021). Standardizing workflows in imaging transcriptomics with the abagen toolbox. Biorxiv.\\n[Q2002N]: Quackenbush, J. (2002). Microarray data normalization and transformation. Nature Genetics, 32(4), 496-501.\\n\""
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Print report content to screen\n",
    "report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "39cfcd88-7b03-491e-b1ba-fb92fcbc4bab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "Regional microarry expression data were obtained from 6 post-mortem brains (1 female, ages 24.0--57.0, 42.50 +/- 13.38) provided by the Allen Human Brain Atlas (AHBA, https://human.brain-map.org; [H2012N]). Data were processed with the abagen toolbox (version 0.1.4+15.gdc4a007; https://github.com/rmarkello/abagen; [M2021B]) using a 68-region surface-based atlas in MNI space.\n",
       "\n",
       "First, microarray probes were reannotated using data provided by [A2019N]; probes not matched to a valid Entrez ID were discarded. Next, probes were filtered based on their expression intensity relative to background noise [Q2002N], such that probes with intensity less than the background in >=50.00% of samples across donors were discarded . When multiple probes indexed the expression of the same gene, we selected and used the probe with the most consistent pattern of regional variation across donors (i.e., differential stability; [H2015N]), calculated with:\n",
       "\n",
       "$$ \\Delta_{{S}}(p) = \\frac{{1}}{{\\binom{{N}}{{2}}}} \\, \\sum_{{i=1}}^{{N-1}} \\sum_{{j=i+1}}^{{N}} \\rho[B_{{i}}(p), B_{{j}}(p)] $$\n",
       "\n",
       "where $ \\rho $ is Spearman's rank correlation of the expression of a single probe, p, across regions in two donors $B_{{i}}$ and $B_{{j}}$, and N is the total number of donors. Here, regions correspond to the structural designations provided in the ontology from the AHBA. \n",
       "\n",
       "The MNI coordinates of tissue samples were updated to those generated via non-linear registration using the Advanced Normalization Tools (ANTs; https://github.com/chrisfilo/alleninf). Samples were assigned to brain regions by minimizing the Euclidean distance between the MNI coordinates of each sample and the nearest surface vertex. Samples where the Euclidean distance to the nearest vertex was more than 2 standard deviations above the mean distance for all samples belonging to that donor were excluded. To reduce the potential for misassignment, sample-to-region matching was constrained by hemisphere and gross structural divisions (i.e., cortex, subcortex/brainstem, and cerebellum, such that e.g., a sample in the left cortex could only be assigned to an atlas parcel in the left cortex; [A2019N]). All tissue samples not assigned to a brain region in the provided atlas were discarded. \n",
       "\n",
       "Inter-subject variation was addressed by normalizing tissue sample expression values across genes using a robust sigmoid function [F2013J]:\n",
       "\n",
       "$$ x_{{norm}} = \\frac{{1}}{{1 + \\exp(-\\frac{{(x-\\langle x \\rangle)}} {{\\text{{IQR}}_{{x}}}})}} $$\n",
       "\n",
       "where $\\langle x \\rangle$ is the median and $\\text{{IQR}}_{{x}}$ is the normalized interquartile range of the expression of a single tissue sample across genes. Normalized expression values were then rescaled to the unit interval: \n",
       "\n",
       "$$ x_{{scaled}} = \\frac{{x_{{norm}} - \\min(x_{{norm}})}} {{\\max(x_{{norm}}) - \\min(x_{{norm}})}} $$\n",
       "\n",
       "Gene expression values were then normalized across tissue samples using an identical procedure. Samples assigned to the same brain region were averaged separately for each donor and then across donors, yielding a regional expression matrix \b.\n",
       "\n",
       "REFERENCES\n",
       "----------\n",
       "[A2019N]: Arnatkevic̆iūtė, A., Fulcher, B. D., & Fornito, A. (2019). A practical guide to linking brain-wide gene expression and neuroimaging data. Neuroimage, 189, 353-367.\n",
       "[F2013J]: Fulcher, B. D., Little, M. A., & Jones, N. S. (2013). Highly comparative time-series analysis: the empirical structure of time series and their methods. Journal of the Royal Society Interface, 10(83), 20130048.\n",
       "[H2012N]: Hawrylycz, M. J., Lein, E. S., Guillozet-Bongaarts, A. L., Shen, E. H., Ng, L., Miller, J. A., ... & Jones, A. R. (2012). An anatomically comprehensive atlas of the adult human brain transcriptome. Nature, 489(7416), 391-399.\n",
       "[H2015N]: Hawrylycz, M., Miller, J. A., Menon, V., Feng, D., Dolbeare, T., Guillozet-Bongaarts, A. L., ... & Lein, E. (2015). Canonical genetic signatures of the adult human brain. Nature Neuroscience, 18(12), 1832.\n",
       "[M2021B]: Markello, R.D., Arnatkevic̆iūtė, A., Poline, J-B., Fulcher, B. D., Fornito, A., & Misic, B. (2021). Standardizing workflows in imaging transcriptomics with the abagen toolbox. Biorxiv.\n",
       "[Q2002N]: Quackenbush, J. (2002). Microarray data normalization and transformation. Nature Genetics, 32(4), 496-501.\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Display report cleanly with LaTeX\n",
    "from IPython.display import display, Markdown\n",
    "display(Markdown(report))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e2f6ccf-f9a4-4737-b53d-1dcc0bb3eab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check what happens when data is re-parcellated to Desikan-Killiany surface atlas, with probe selection parameter changed to be mean max probe intensity \n",
    "#across donors instead of default setting, and generate report at the same time\n",
    "expression_maxprobe, report_maxprobe = abagen.get_expression_data(atlas['image'], atlas['info'], probe_selection='max_intensity', return_report=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34060b5e-2cdf-4034-8e5e-baba6be50f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Print to screen contents of newly generated expression_maxprobe (panda.dataframe) object \n",
    "expression_maxprobe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7af3bd47-7c73-4c9d-9b6e-1e9095d00b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Display report_maxprobe cleanly with LaTeX\n",
    "from IPython.display import display, Markdown\n",
    "display(Markdown(report_maxprobe))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0d463d9-ffcc-48c1-8a61-d2e886d31a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save expression pandas.dataframe as csv file for downstream analysis\n",
    "expression.to_csv('Expression_AHBA_abagen_defaultparam.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2299a378-64e6-471f-b740-625a1da140c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save expression_maxprobe pandas.dataframe as csv file for downstream analysis\n",
    "expression_maxprobe.to_csv('Expression_AHBA_abagen_maxprobe.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35adcc4e-c9ce-4d0b-af45-d11d9f3ca594",
   "metadata": {},
   "outputs": [],
   "source": [
    "#View info on DK atlas regions and copy to home directory\n",
    "!head /srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/data/atlas-desikankilliany.csv\n",
    "!tail /srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/data/atlas-desikankilliany.csv\n",
    "!cp /srv/conda/envs/notebook/lib/python3.10/site-packages/abagen/data/atlas-desikankilliany.csv ./atlas-desikankilliany_copy.csv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
